### PR TITLE
Adding bindings for D2D1, DWrite, DXGI, and some of Windowscodecs

### PR DESCRIPTION
**Pull request by Connorcpu**
This pull request adds interface bindings for D2D1, DWrite, DXGI, and some of Windowscodecs (enough that interfaces the other libraries use are defined).

### CODE REVIEW

File: lib/d2d1-sys/build.rs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // Copyright © 2015, Peter Atashian
 // Licensed under the MIT License <LICENSE.md>
 fn main() {
-    println!("cargo:rustc-flags=-l d2d1");
+    println!("cargo:rustc-link-lib=dylib=d2d1");
```
Comment (retep998) [inline lib/d2d1-sys/build.rs]: Just switch this over to using `winapi-build`.

File: lib/dxgi-sys/src/lib.rs [hunk 2]
```diff
@@ -16,3 +16,20 @@ extern "system" {
     // pub fn DXGIGetDebugInterface1();
     // pub fn DXGIReportAdapterConfiguration();
 }
+#[cfg(any(target_arch = "x86", target_arch = "x86_64", target_arch = "arm"))]
+extern "C" {
+    pub static IID_IDXGIObject: IID;
```
Comment (retep998) [inline lib/dxgi-sys/src/lib.rs]: These are provided by `dxguid`, not `dxgi`.

File: src/d2d1.rs [hunk 3]
```diff
@@ -0,0 +1,1405 @@
+// Copyright © 2015, Connor Hilarides
+// Licensed under the MIT License <LICENSE.md>
+//! Mappings for the contents of d2d1.h, d2dbasetypes.h, and dcommon.h
+
+//-------------------------------------------------------------------------------------------------
+// D2D1_DEFAULT_FLATTENING_TOLERANCE
+//-------------------------------------------------------------------------------------------------
```
Comment (retep998) [inline src/d2d1.rs]: Excessive comments everywhere. Think of all the wasted vertical space!

File: src/d2d1.rs [hunk 4]
```diff
@@ -0,0 +1,1405 @@
...
+    fn GetColorInterpolationGamma(&mut self) -> D2D1_GAMMA,
+    fn GetExtendMode(&mut self) -> D2D1_EXTEND_MODE
+});
+
+//-------------------------------------------------------------------------------------------------
+// ID2D1Brush
+//-------------------------------------------------------------------------------------------------
+RIDL!(
+interface ID2D1Brush(ID2D1BrushVtbl): ID2D1Resource(ID2D1ResourceVtbl) {
+    fn SetOpacity(&mut self, opacity: ::FLOAT) -> (),
```
Comment (retep998) [inline src/d2d1.rs]: It feels weird seeing `-> ()`, but then I remember that `RIDL!` requires a return type be specified.

File: src/d2d1.rs (hunk 3 above)
Comment (Connorcpu) [inline src/d2d1.rs]: Heh, I was just trying to provide some separation between the walls of constants, should I remove them?

File: src/dxgi.rs [hunk 5]
```diff
@@ -0,0 +1,683 @@
...
+    pub DeviceName: [::WCHAR; 32],
+    pub DesktopCoordinates: ::RECT,
+    pub AttachedToDesktop: ::BOOL,
+    pub Rotation: DXGI_MODE_ROTATION,
+    pub Monitor: ::HMONITOR,
+}
+
+impl Clone for DXGI_OUTPUT_DESC {
+    fn clone(&self) -> DXGI_OUTPUT_DESC {
+        DXGI_OUTPUT_DESC {
```
Comment (retep998) [inline src/dxgi.rs]: Just return `*self` since the type is `Copy`.

File: src/dxgi.rs [hunk 6]
```diff
@@ -0,0 +1,683 @@
...
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct DXGI_SHARED_RESOURCE {
+    pub Handle: ::HANDLE,
+}
+
+//-------------------------------------------------------------------------------------------------
+// DXGI_RESOURCE_PRIORITY
+//-------------------------------------------------------------------------------------------------
+pub const DXGI_RESOURCE_PRIORITY_MINIMUM: ::DWORD =  ( 0x28000000 );
```
Comment (retep998) [inline src/dxgi.rs]: Aligning like this is a no-no.

**Comment by Connorcpu** [comment]
Is there anything else I should do for this?

File: lib/windowscodecs-sys/src/lib.rs [hunk 7]
```diff
@@ -6,4 +6,4 @@
 extern crate winapi;
 use winapi::*;
 extern "system" {
-}
+}
```
Comment (retep998) [inline lib/windowscodecs-sys/src/lib.rs]: Why did you remove the newline?

File: src/d2d1.rs [hunk 8]
```diff
@@ -0,0 +1,1137 @@
...
+// Licensed under the MIT License <LICENSE.md>
+//! Mappings for the contents of d2d1.h, d2dbasetypes.h, and dcommon.h
+
+use std::ops::Mul;
+use std::default::Default;
+
+pub const D2D1_DEFAULT_FLATTENING_TOLERANCE: ::FLOAT = 0.25;
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
```
Comment (retep998) [inline src/d2d1.rs]: Could you put the `#[repr]` and `#[derive]` on one line everywhere to save space?

File: src/d2d1.rs [hunk 9]
```diff
@@ -0,0 +1,1137 @@
...
+pub struct D2D_MATRIX_3X2_F {
+    pub _11: ::FLOAT,
+    pub _12: ::FLOAT,
+    pub _21: ::FLOAT,
+    pub _22: ::FLOAT,
+    pub _31: ::FLOAT,
+    pub _32: ::FLOAT,
+}
+
+impl Mul for D2D_MATRIX_3X2_F {
```
Comment (retep998) [inline src/d2d1.rs]: At the moment the only traits that should be implemented are `Copy` `Clone` and `Debug`. I'd prefer not have things like `Mul` or `PartialEq` at the moment.

File: src/d3dcommon.rs [hunk 10]
```diff
@@ -0,0 +1,630 @@
+// Copyright © 2015; Connor Hilarides
+// Licensed under the MIT License <LICENSE.md>
+//! Mappings for the contents of d3dcommon.h
+
+pub type D3D_DRIVER_TYPE = ::DWORD;
+pub const D3D_DRIVER_TYPE_UNKNOWN: ::DWORD = 0;
+pub const D3D_DRIVER_TYPE_HARDWARE: ::DWORD = (D3D_DRIVER_TYPE_UNKNOWN + 1);
```
Comment (retep998) [inline src/d3dcommon.rs]: No need for extra parentheses.

File: src/d3dcommon.rs [hunk 11]
```diff
@@ -0,0 +1,630 @@
...
+pub type D3D_DRIVER_TYPE = ::DWORD;
+pub const D3D_DRIVER_TYPE_UNKNOWN: ::DWORD = 0;
+pub const D3D_DRIVER_TYPE_HARDWARE: ::DWORD = (D3D_DRIVER_TYPE_UNKNOWN + 1);
+pub const D3D_DRIVER_TYPE_REFERENCE: ::DWORD = (D3D_DRIVER_TYPE_HARDWARE + 1);
+pub const D3D_DRIVER_TYPE_NULL: ::DWORD = (D3D_DRIVER_TYPE_REFERENCE + 1);
+pub const D3D_DRIVER_TYPE_SOFTWARE: ::DWORD = (D3D_DRIVER_TYPE_NULL + 1);
+pub const D3D_DRIVER_TYPE_WARP: ::DWORD = (D3D_DRIVER_TYPE_SOFTWARE + 1);
+
+pub type D3D_FEATURE_LEVEL = ::DWORD;
+pub const D3D_FEATURE_LEVEL_9_1: ::DWORD   = 0x9100;
```
Comment (retep998) [inline src/d3dcommon.rs]: Unnecessary vertical alignment.

File: src/d3dcommon.rs [hunk 12]
```diff
@@ -0,0 +1,630 @@
...
+pub const D3D11_PRIMITIVE_TOPOLOGY_POINTLIST: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_POINTLIST;
+pub const D3D11_PRIMITIVE_TOPOLOGY_LINELIST: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_LINELIST;
+pub const D3D11_PRIMITIVE_TOPOLOGY_LINESTRIP: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_LINESTRIP;
+pub const D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST;
+pub const D3D11_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP;
+pub const D3D11_PRIMITIVE_TOPOLOGY_LINELIST_ADJ: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_LINELIST_ADJ;
+pub const D3D11_PRIMITIVE_TOPOLOGY_LINESTRIP_ADJ: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_LINESTRIP_ADJ;
+pub const D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST_ADJ: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST_ADJ;
+pub const D3D11_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP_ADJ: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP_ADJ;
+pub const D3D11_PRIMITIVE_TOPOLOGY_1_CONTROL_POINT_PATCHLIST: ::DWORD = D3D_PRIMITIVE_TOPOLOGY_1_CONTROL_POINT_PATCHLIST;
```
Comment (retep998) [inline src/d3dcommon.rs]: Use some newlines here to keep within 99 columns.

File: src/d3dcommon.rs [hunk 13]
```diff
@@ -0,0 +1,630 @@
...
+pub type _D3D_SHADER_VARIABLE_CLASS = ::DWORD;
+pub type D3D_SHADER_VARIABLE_CLASS = ::DWORD;
+pub const D3D_SVC_SCALAR: ::DWORD = 0;
+pub const D3D_SVC_VECTOR: ::DWORD = ( D3D_SVC_SCALAR + 1 );
+pub const D3D_SVC_MATRIX_ROWS: ::DWORD = ( D3D_SVC_VECTOR + 1 );
+pub const D3D_SVC_MATRIX_COLUMNS: ::DWORD = ( D3D_SVC_MATRIX_ROWS + 1 );
+pub const D3D_SVC_OBJECT: ::DWORD = ( D3D_SVC_MATRIX_COLUMNS + 1 );
+pub const D3D_SVC_STRUCT: ::DWORD = ( D3D_SVC_OBJECT + 1 );
+pub const D3D_SVC_INTERFACE_CLASS: ::DWORD = ( D3D_SVC_STRUCT + 1 );
+pub const D3D_SVC_INTERFACE_POINTER: ::DWORD = ( D3D_SVC_INTERFACE_CLASS + 1 );
```
Comment (retep998) [inline src/d3dcommon.rs]: More extra parentheses and weird inside spacing.

File: src/d3dcommon.rs [hunk 14]
```diff
@@ -0,0 +1,630 @@
...
+pub const D3D_SIT_TEXTURE: ::DWORD = ( D3D_SIT_TBUFFER + 1 ) ;
+pub const D3D_SIT_SAMPLER: ::DWORD = ( D3D_SIT_TEXTURE + 1 ) ;
+pub const D3D_SIT_UAV_RWTYPED: ::DWORD = ( D3D_SIT_SAMPLER + 1 ) ;
+pub const D3D_SIT_STRUCTURED: ::DWORD = ( D3D_SIT_UAV_RWTYPED + 1 ) ;
+pub const D3D_SIT_UAV_RWSTRUCTURED: ::DWORD = ( D3D_SIT_STRUCTURED + 1 ) ;
+pub const D3D_SIT_BYTEADDRESS: ::DWORD = ( D3D_SIT_UAV_RWSTRUCTURED + 1 ) ;
+pub const D3D_SIT_UAV_RWBYTEADDRESS: ::DWORD = ( D3D_SIT_BYTEADDRESS + 1 ) ;
+pub const D3D_SIT_UAV_APPEND_STRUCTURED: ::DWORD = ( D3D_SIT_UAV_RWBYTEADDRESS + 1 ) ;
+pub const D3D_SIT_UAV_CONSUME_STRUCTURED: ::DWORD = ( D3D_SIT_UAV_APPEND_STRUCTURED + 1 ) ;
+pub const D3D_SIT_UAV_RWSTRUCTURED_WITH_COUNTER: ::DWORD = ( D3D_SIT_UAV_CONSUME_STRUCTURED + 1 ) ;
```
Comment (retep998) [inline src/d3dcommon.rs]: I am of the opinion that these parentheses with spaces inside look gross.

File: src/windowscodecs.rs [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+pub const WICPngHistFrequencies: ::DWORD = 0x1;
+
+pub type WICPngIccpProperties = ::DWORD;
+pub const WICPngIccpProfileName: ::DWORD = 0x1;
+pub const WICPngIccpProfileData: ::DWORD = 0x2;
+
+pub type WICPngSrgbProperties = ::DWORD;
+pub const WICPngSrgbRenderingIntent: ::DWORD = 0x1;
+
+pub type WICPngTimeProperties = ::DWORD;
```
Comment (retep998) [inline src/windowscodecs.rs]: Shouldn't this and the other things here be enums?

Comment (Connorcpu) [inline src/windowscodecs.rs]: I was doing all of those like that to be consistent with other parts of winapi-rs, but I'm happy to redo them as enums.
